amazon/ion/ioncmodule.c

```diff
@@ -0,0 +1,1369 @@
+ *
+ */
+static iERR ionc_write_sequence(hWRITER writer, PyObject* sequence, PyObject* tuple_as_sexp) {
+    iENTER;
+    PyObject* child_obj = NULL;
+    sequence = PySequence_Fast(sequence, "expected sequence");
+    Py_ssize_t len = PySequence_Size(sequence);
+    Py_ssize_t i;
+
+    for (i = 0; i < len; i++) {
+        child_obj = PySequence_Fast_GET_ITEM(sequence, i);
+        Py_INCREF(child_obj);
+
+        IONCHECK(Py_EnterRecursiveCall(" while writing an Ion sequence"));
+        err = ionc_write_value(writer, child_obj, tuple_as_sexp);
+        Py_LeaveRecursiveCall();
+        IONCHECK(err);
+
+        Py_DECREF(child_obj);
+        child_obj = NULL;
+    }
+fail:
+    Py_XDECREF(child_obj);
+    Py_DECREF(sequence);
+    cRETURN;
+}
+
+/*
+ *  Writes a struct
+ *
+ *  Args:
+ *      writer:  An ion writer
+ *      map: An ion python struct
+ *      tuple_as_sexp: Decides if a tuple is treated as sexp
+ *
+ */
+static iERR ionc_write_struct(hWRITER writer, PyObject* map, PyObject* tuple_as_sexp) {
+    iENTER;
+    PyObject * list = PyMapping_Items(map);
+    PyObject * seq = PySequence_Fast(list, "expected a sequence within the map.");
+    PyObject * key = NULL, *val = NULL, *child_obj = NULL;
+    Py_ssize_t len = PySequence_Size(seq);
+    Py_ssize_t i;
+
+    for (i = 0; i < len; i++) {
+        child_obj = PySequence_Fast_GET_ITEM(seq, i);
+        key = PyTuple_GetItem(child_obj, 0);
+        val = PyTuple_GetItem(child_obj, 1);
+        Py_INCREF(child_obj);
+        Py_INCREF(key);
+        Py_INCREF(val);
+
+        if (PyUnicode_Check(key)) {
+            ION_STRING field_name;
+            ion_string_from_py(key, &field_name);
+            IONCHECK(ion_writer_write_field_name(writer, &field_name));
+        }
+        else if (key == Py_None) {
+            // if field_name is None, write symbol $0 instead.
+            IONCHECK(_ion_writer_write_field_sid_helper(writer, 0));
+        }
+
+        IONCHECK(Py_EnterRecursiveCall(" while writing an Ion struct"));
+        err = ionc_write_value(writer, val, tuple_as_sexp);
+        Py_LeaveRecursiveCall();
+        IONCHECK(err);
+
+        Py_DECREF(child_obj);
+        Py_DECREF(key);
+        Py_DECREF(val);
+        child_obj = NULL;
+        key = NULL;
+        val = NULL;
+    }
+    Py_XDECREF(list);
+    Py_XDECREF(seq);
+fail:
+    Py_XDECREF(child_obj);
+    Py_XDECREF(key);
+    Py_XDECREF(val);
+    cRETURN;
+}
+
+/*
+ *  Writes an int
+ *
+ *  Args:
+ *      writer:  An ion writer
+ *      obj: An ion python int
+ *
+ */
+static iERR ionc_write_big_int(hWRITER writer, PyObject *obj) {
+    iENTER;
+    PyObject* int_str = PyObject_CallMethod(obj, "__str__", NULL);
+    ION_STRING string_value;
+    ion_string_from_py(int_str, &string_value);
+    ION_INT ion_int_value;
+
+    IONCHECK(ion_int_init(&ion_int_value, NULL));
+    IONCHECK(ion_int_from_string(&ion_int_value, &string_value));
+    IONCHECK(ion_writer_write_ion_int(writer, &ion_int_value));
```

Comment:
I implement converting between arbitrary python int and ion int with math first because converting without string is a little bit tricky. (e.g. need to take care of max int overflow and unsigned ion_int digits, some python math issue such as 18446744073709551615 = (2^62)*3 +1 but 18446744073709551615/(2^62) = 4 and so on..)  

After making sure this change is what we want, I'll implement the rest of "avoiding string" part (decimal, float and so on.).

Write(dumps) code is [here](https://github.com/cheqianh/ion-python/compare/c_ext_final_review...cheqianh:change#diff-535412a8258ff8b32325ed24da1018a8b331ec818afbda63f3d585d631c5847dR461)
Read(loads) code is [here](https://github.com/cheqianh/ion-python/compare/c_ext_final_review...cheqianh:change#diff-535412a8258ff8b32325ed24da1018a8b331ec818afbda63f3d585d631c5847dR1083)

Reply:
Looks good to me.